### ISSUE TITLE
The hooks use a hard-coded value for GRID P40-6Q.

### ISSUE THREAD

**Issue by DualCoder** [DualCoder/vgpu_unlock#2]
There is a single hard-coded value on line 804 of `vgpu_unlock_hooks.c` that makes the resulting lookup table match the `GRID P40-6Q` device:
```c
pci_info[3] = 0x11ec;
```

This value is printed next to the PCI device ID in the list of supported vGPUs printed by `nvidia-vgpud`:
```
nvidia-vgpud[674]: Supported VGPU 0x32: max 4
nvidia-vgpud[674]: VGPU Type 0x32: GRID P40-6Q Class: Quadro
nvidia-vgpud[674]: DevId: 0x10de / 0x1b38 / 0x10de / 0x11ec
nvidia-vgpud[674]: Framebuffer: 0x164000000
nvidia-vgpud[674]: Mappable video size: 0x400000
nvidia-vgpud[674]: Framebuffer reservation: 0x1c000000
nvidia-vgpud[674]: FRL configuration: 0x3c
nvidia-vgpud[674]: CUDA enabled: 0x1
nvidia-vgpud[674]: ECC supported: 0x1
nvidia-vgpud[674]: Multi vGPU supported: 0x0
nvidia-vgpud[674]: Encoder Capacity: 0x64
nvidia-vgpud[674]: BAR1 Length: 0x100
nvidia-vgpud[674]: Frame Rate Limiter enabled: 0x1
nvidia-vgpud[674]: Number of Displays: 4
nvidia-vgpud[674]: Max pixels: 58982400
nvidia-vgpud[674]: Display: width 7680, height 4320
nvidia-vgpud[674]: License: Quadro-Virtual-DWS,5.0;GRID-Virtual-WS,2.0;GRID-Virtual-WS-Ext,2.0
```
As a result vgpu\_unlock will only support the `GRID P40-6Q` device. We will need to have a lookup table of all these device IDs per GPU so that they can all be inserted into the lookup tables in the kernel module.

A workaround is to edit the hard-coded value and then remove and reinstall the kernel module using dkms.

**Comment by KrutavShah** [comment]
This is perfect! Assuming that only one profile is currently working, the `GRID P40-6Q` device, it's still great progress and we can explore getting the others to work. Would creating a script to unload the kernel module, modify the necessary values, and then load it back in be a feasible solution for this bug?

**Comment by DualCoder** [comment]
> This is perfect! Assuming that only one profile is currently working, the `GRID P40-6Q` device, it's still great progress and we can explore getting the others to work. Would creating a script to unload the kernel module, modify the necessary values, and then load it back in be a feasible solution for this bug?

That could possibly work, but it seems a lot more complicated than just adding all the IDs into `vgpu_unlock_hooks.c`. And with that solution you would only be able to use a single profile at a time.

> Edit: I noticed that modifying any of those values like the ones you listed above, such as "License:" or "ECC supported:" seems to disable the specified VGPU configuration, making it not possible to use that profile in the hypervisor. Tested on Red Hat and GTX 1080 spoofed to Tesla P4.

Could you clarify? What did you modify and what step failed when you tried to use the profile?

**Comment by KrutavShah** [comment]
> That could possibly work, but it seems a lot more complicated than just adding all the IDs into vgpu_unlock_hooks.c. And with that solution you would only be able to use a single profile at a time. 

If I understand the vGPU documentation correctly, you only actually can use one profile at a time on a card. For example, 6 2GB Quadro profiles on one GPU.

> Could you clarify? What did you modify and what step failed when you tried to use the profile?

I modified the values `nvidia-vgpud[674]: ECC supported: 0x1` and `nvidia-vgpud[674]: License: Quadro-Virtual-DWS` and what I found is that modifying any of these values seems to make the profile inaccessible. In other words, the GUI for the hypervisor provides a convenient drop-down menu for the profiles on the Tesla P4. For example, I modified a value on the P4-4Q profile and that profile now no longer shows up in the menu.

Some more digging reveals that these profiles are backed by signatures for each of their values. I don't know what's the workaround for that just yet, but one thing I can say is that if the end user of this script wants to use a particular profile like P40-6Q, there could be a way for the user to modify the script for that one particular profile, even if it doesn't let them use a different profile. For example, someone with a 12GB GP102 card could run two of those 6Q profiles, but not one 6Q and 3 2Q profiles at the same time without vGPU rejecting the configuration in the first place.
